docs/walkthrough.md

```diff
@@ -0,0 +1,320 @@
+
+# Table of Contents
+- [How to Use This Guide](#how-to-use-this-guide)
+  - [Gotchas](#gotchas)
+- [Goals for the Environment](#goals-for-the-environment)
+  - [Q&A](#qa)
+- [Your Platform Setup](#your-platform-setup)
+  - [Global Dependencies](#global-dependencies)
+- [Workflow 1: Locking a New Project](#workflow-1-locking-a-new-project)
+- [Workflow 2: Locking an Existing Project (TBD)](#workflow-2-locking-an-existing-project)
+- [Reproduce](#reproduce)
+  - [With Docker in Mind](#with-docker-in-mind)
+- [Upgrade](#upgrade)
+  - [Upgrade an existing package](#upgrade-an-existing-package)
+  - [Adding a New Package via `conda`](#adding-a-new-package-via-conda)
+  - [Adding a New Package via `pip`](#adding-a-new-package-via-pip)
+  - [A Quick Note on Downgrading](#a-quick-note-on-downgrading)
+- [Debugging](#debugging)
+  - [`conda-lock` giving vague `AssertionError`](#conda-lock-giving-vague-assertionerror)
+- [Terms](#terms)
+- [References and Resources](#references-and-resources)
+- [Versions](#versions)
+
+# How to Use This Guide
+This is a guide on how to produce a reproducible and upgradeable environment for Python applications. A great deal of inspiration came from a pythonspeed.com article about this very topic.<sup>[4](#references-and-resources)</sup> While the article is a wonderful resource, I needed more so I developed this guide.
+
+You can use the table of contents to get a quick look at topics. Or you can skip ahead to [Your Platform Setup](#your-platform-setup) which is the beginning of the walkthrough.
+
+Ultimately, this guide is just that, a *guide*. The goal was to make the concepts easy to follow so that you can apply it to your use case. You may use a different platform than this guide but hopefully still benefit from the content.
+
+## Gotchas
+* I use `pipx` for parts of this guide, you may choose not to.
+* My platform may not be your platform so keep an eye out for incompatible commands.
+* For simplicity, the guide assumes no explicit versions of packages.
+# Goals for the Environment
+**Reproducible**: The environment can be reproduced across platforms.
+**Upgradeable**: Packages in the environment can be easily changed (e.g., version, new package, etc.)
+
+## Q&A
+
+### Why do you suggest using  `conda`  AND  `pip`?
+Sometimes a package installed via `conda` (even using the `conda-forge` channel) breaks my environment. I have found using `pip` as the *exception to the rule* has resolved most of these issues.
+
+### Why is this process slow?
+Solving the dependencies of dependencies for one operating system is time consuming. Out of the box,  `conda-lock`  will try to solve for 3 different operating systems (i.e., linux-64, osx-64, and win-64). However, `conda` recently announced partnering with a `mamba` solver. You could configure your `conda` instance to use a faster solver like this one. This is beyond the scope of this walkthrough. Regardless, it will still take time.
```

Comment:
`conda-lock` defaults to using mamba if its installed so we should just recommend that.

Reply:
I have been using `miniconda` with the `libmamba` solver enabled ([ref](https://www.anaconda.com/blog/a-faster-conda-for-a-growing-community)). Is that much different from using `mamba` directly?

Reply:
they're similar but not the same

Reply:
Not trying to be obtuse here, but wouldn't a strong dependency like this discourage a variety of users from engaging `conda-lock`? I am happy to adapt this documentation, but wanted to approach this philosophical point first.

For example, I could make the point in the documentation so readers are not misled into thinking it only works with the `mamba` setup.